Adagrad, L2 regularization, log-loss, and beam-search parsing

cgpotts opened the pull request:
Hey @reschkek,

This PR is for core optimization improvements to SippyCup. I managed to do everything I wanted, but this is really straining SippyCup, which has baked in a lot of stuff that should be abstracted out, and its inconsistent use of object-orientation makes everything hard to keep track of. I press on:

The core task was to allow beam search during parsing. These changes are pretty straightforward. They are in `parsing.py`. The new pieces are just that `parse_input` and its subfunctions take an optional `scorer` parameter. This needs to map `Parse` objects to floats. In turn, `Grammar.parse_input` also takes a `scorer` parameter, which is just passed to the module `parse_input`.

At present, the beam size is determined by `parsing.MAX_CELL_CAPACITY`. This is opportunistic, but it works, and I couldn't bear to add more parameters.

I also updated `learning.py`. `latent_sgd` now supports the log loss as well as the old hinge loss. I also added L2 regularization and AdaGrad. AdaGrad seems to help a ton. I also wove beam_search into this function.

All of this is needlessly complicated, but it works. With all the old default parameters, the behavior should be exactly as it was before I made these changes, with one small exception: if beam search is not used during parsing, the cells get filled with a random sample of the parses for that span. Before, it was like the first `parsing.MAX_CELL_CAPACITY` parses, which is impossible to justify, as far as I can tell.

I'll shortly check in some notebooks to `suggestion_engine.develop` that provide some tests and examples of how this all fits together.

cgpotts committed:
Log-loss, L2 penalty, AdaGrad, and beam search scoring

cgpotts committed:
Beam search parsing

cgpotts committed:
Passing an optional scorer to the grammar for beam search

reschkek commented:
Thanks!  I'll merge this after I've had a chance to try it out with the notebooks you are producing.

One comment:

You wrote:

```
...if beam search is not used during parsing, the cells get filled with a random sample of the parses for that span. Before, it was like the first parsing.MAX_CELL_CAPACITY parses, which is impossible to justify, as far as I can tell"
```

I can justify using the first MAX_CELL_CAPACITY parses instead of a random sample.  Doing so ensures deterministic parsing--given an input, the parser produces the same output every time it runs.  This is valuable for debugging and for automated evaluation.

In general, in systems that require randomization, we can still enable reproducibility (ie determinism) by exposing a random seed that the user can set.

cgpotts commented:
@reschkek To ensure deterministic behavior, I would favor setting the random seed, which is supported by the training method, and we could add support elsewhere too. Taking the first `n` parses for each cell creates a dependency on the order in which the grammar rules were added. That's the part I can't justify.

cgpotts commented:
@reschkek I meant to add: this is a notebook for testing/exploring the new functionality: 

https://github.com/roaminsight/suggestion-engine/blob/master/develop/2016-11-13-cgp-beam-search/alien_math.ipynb

reschkek committed:
Use package-prefixed imports

This allows users to use sippycup as a package without
putting sippycup itself on the PYTHONPATH

reschkek commented:
@cgpotts 👍 This is great. I was able to run your example notebook.

I'm going to merge this as is.  Later, if we wish to have more control over the random seed beyond setting it in the training method, we can add this functionality later.

cgpotts commented:
@reschkek At some point soon, let's take a few minutes to talk about how I can make the desired kind of scoring easier for you (and perhaps how we might make SippyCup easier to work with).
